data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/SqsDynamicVisibilityTimeoutManager.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.source;
+
+import org.opensearch.dataprepper.plugins.source.configuration.SqsOptions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.sqs.SqsClient;
+import software.amazon.awssdk.services.sqs.model.ChangeMessageVisibilityRequest;
+import software.amazon.awssdk.services.sqs.model.Message;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class SqsDynamicVisibilityTimeoutManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SqsDynamicVisibilityTimeoutManager.class);
+    static final int VISIBILITY_TIMEOUT_SECONDS_PER_QUARTER_GB = 30;
+    static final int QUARTER_GB_BYTES = 250_000_000;
+    static final int MAX_VISIBILITY_TIMEOUT_SECONDS = 43_200;
+    static final int MINIMUM_WAIT_BEFORE_CHANGE_VISIBILITY_TIMEOUT_SECONDS = 2;
+    static final int TIME_BETWEEN_CHANGE_MESSAGE_CALL_AND_VISIBILITY_TIMEOUT_SECONDS = 10;
+
+    private final SqsClient sqsClient;
+    private final ScheduledExecutorService scheduledExecutorService;
+    private final String queueUrl;
+    private final int initialMessageVisibilityTimeout;
+    private final int initialWaitBeforeIncrease;
+
+    private final Map<String, ScheduledFuture<?>> messageToScheduledFuture;
+
+    public SqsDynamicVisibilityTimeoutManager(final SqsClient sqsClient,
+                                              final SqsOptions sqsOptions) {
+        this.sqsClient = sqsClient;
+        this.scheduledExecutorService = Executors.newScheduledThreadPool(sqsOptions.getMaximumMessages());
+        this.queueUrl = sqsOptions.getSqsUrl();
+        this.messageToScheduledFuture = new HashMap<>();
+        this.initialMessageVisibilityTimeout = (int) sqsOptions.getVisibilityTimeout().getSeconds();
+        this.initialWaitBeforeIncrease = Math.max(MINIMUM_WAIT_BEFORE_CHANGE_VISIBILITY_TIMEOUT_SECONDS,
+                initialMessageVisibilityTimeout - TIME_BETWEEN_CHANGE_MESSAGE_CALL_AND_VISIBILITY_TIMEOUT_SECONDS);
+    }
+
+
+
+    public void startDynamicVisibilityForMessage(final Message message, final long objectSizeBytes) {
+        final int visibilityTimeoutIncrease = calculateVisibilityTimeoutFromObjectSize(objectSizeBytes);
+        final int waitBeforeIncreasingAgain = Math.min(MINIMUM_WAIT_BEFORE_CHANGE_VISIBILITY_TIMEOUT_SECONDS,
+                visibilityTimeoutIncrease - TIME_BETWEEN_CHANGE_MESSAGE_CALL_AND_VISIBILITY_TIMEOUT_SECONDS);
+
+        final ScheduledFuture<?> scheduledFuture = scheduledExecutorService.scheduleAtFixedRate(
```

Comment:
If we hit an error, it seems that scheduling can become stuck. So I think it may advantageous to set the source of truth for in-flight processing to be in the form of a callback from the S3Service (and thus the S3ObjectWorker) itself.

Right now, there is already a callback from the `InputCodec` back to the different S3 workers. `S3ObjectWorker` already has a field:

```
private final BiConsumer<Event, S3ObjectReference> eventConsumer;
```

We can add a "contextual consumer" which is passed in. (We might even be able to make this more static and mapped from the S3ObjectReference)

```
public void parseS3Object(final S3ObjectReference s3ObjectReference, final AcknowledgementSet acknowledgementSet, BiConsumer<Event, S3ObjectReference> contextualConsumer) {
  ...
  final BiConsumer<Event, S3ObjectReference> eventConsumerToUse = contextualConsumer.andThen(eventConsumer);
  ...
}
```

Use it in the codec:

```
codec.parse(inputStream, record -> {
                    // relevant change
                    contextualConsumer.accept(record.getData(), s3ObjectReference);
                    ...
            });
```

The SQS code can provide this contextual consumer. This can then report back some factual statement as to whether the message is still in-flight or not.